### PR TITLE
SCT-479 - Create "Delete Relationship" Service API Endpoint

### PR DESCRIPTION
**Pull request by mapleleafjack**
## Link to JIRA ticket

[Ticket link](https://hackney.atlassian.net/browse/SCT-479)

### *What is the problem we're trying to solve*

Added a DELETE endpoint that deletes saved relationship.

### *What changes have we introduced*

+ Introduced a new endpoint `relationships/personal/{id}` that deletes the relationship with matching ID
+ Added a new "getRelationshipByID" function in the DatabaseGateway
+ Added a new "DeleteRelationships" function in the DatabaseGateway that deletes the passed relationship, its reverse row and the linked relationship details.
+ Added new helper function SavePersonWithPersonalRelationshipAndOppositeToDatabase
+ Added a sql update script in `database/schema.sql` that updates the sccv_personal_relationship_type sequence ID in the test DB (other changes are only whitespace fixes by the linter)
+ Tests for getRelationshipByID and DeleteRelationships

#### _Checklist_

- [ ] Added end-to-end (i.e. integration) tests where necessary e.g to test complete functionality of a newly added feature
- [X] Added tests to cover all new production code
- [ ] Added comments to the README or updated relevant documentation _(add link to documentation)_, where necessary.
- [X] Checked all code for possible refactoring
- [X] Code pipeline builds correctly

### *Follow up actions after merging PR*

Aligning the FE so the delete relationship calls this endpoint

### CODE REVIEW

File: SocialCareCaseViewerApi.Tests/V1/Gateways/Database/GetPersonalRelationshipByIdTests.cs [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+        {
+            _mockSystemTime = new Mock<ISystemTime>();
+            _databaseGateway = new DatabaseGateway(DatabaseContext, _mockProcessDataGateway.Object, _mockSystemTime.Object);
+            DatabaseContext.ChangeTracker.QueryTrackingBehavior = QueryTrackingBehavior.NoTracking;
+        }
+
+        [Test]
+        public void WhenNoMatchingIDReturnsNull()
+        {
+            var response = _databaseGateway.GetPersonWithPersonalRelationshipsByPersonId(123456789);
```
Comment (chubberlisk) [inline SocialCareCaseViewerApi.Tests/V1/Gateways/Database/GetPersonalRelationshipByIdTests.cs:30]:
```suggestion
            var response = _databaseGateway.GetPersonalRelationshipById(123456789);
```

File: SocialCareCaseViewerApi/V1/Controllers/RelationshipController.cs [hunk 2]
```diff
@@ -79,5 +79,33 @@ public IActionResult CreatePersonalRelationship([FromBody] CreatePersonalRelatio
...
+        /// <summary>
+        /// Delete a personal relationship
+        /// </summary>
+        /// <param name="id"></param>
+        /// <response code="200">Successfully removed a relationship.</response>
+        /// <response code="400">Invalid DeletePersonalRelationshipRequest received</response>
+        [ProducesResponseType(StatusCodes.Status200OK)]
+        [HttpDelete]
+        [Route("relationships/personal/{id:long}")]
+        public IActionResult RemovePersonalRelationshipRequest(long id)
```
Comment (chubberlisk) [inline SocialCareCaseViewerApi/V1/Controllers/RelationshipController.cs:92]:
```suggestion
        public IActionResult RemovePersonalRelationship(long id)
```

Comment (chubberlisk) [reply SocialCareCaseViewerApi/V1/Controllers/RelationshipController.cs:92]: Could we add controller tests for this endpoint? See https://github.com/LBHackney-IT/social-care-case-viewer-api/tree/development/SocialCareCaseViewerApi.Tests/V1/Controllers/Relationship for examples.

File: SocialCareCaseViewerApi/V1/Exceptions/CustomExceptions.cs [hunk 3]
```diff
@@ -63,6 +63,11 @@ public class PersonNotFoundException : Exception
         public PersonNotFoundException(string message) : base(message) { }
     }
 
+    public class RelationshipNotFoundException : Exception
```
Comment (chubberlisk) [inline SocialCareCaseViewerApi/V1/Exceptions/CustomExceptions.cs:66]: What do you think about renaming this to `PersonalRelationshipNotFoundException` as we have other `PersonalRelationship` exceptions below.

File: SocialCareCaseViewerApi/V1/Gateways/DatabaseGateway.cs [hunk 4]
```diff
@@ -927,6 +927,38 @@ public PersonalRelationshipType GetPersonalRelationshipTypeByDescription(string
                 .FirstOrDefault(prt => prt.Description.ToLower() == description.ToLower());
         }
 
+        public Infrastructure.PersonalRelationship GetPersonalRelationshipById(long relationshipId)
+        {
+            return _databaseContext.PersonalRelationships
+                .FirstOrDefault(prt => prt.Id == relationshipId);
+        }
+
+        public void DeleteRelationships(Infrastructure.PersonalRelationship relationship)
```
Comment (chubberlisk) [inline SocialCareCaseViewerApi/V1/Gateways/DatabaseGateway.cs:936]: Looks like where missing tests for this method?

File: SocialCareCaseViewerApi/V1/Gateways/DatabaseGateway.cs [hunk 5]
```diff
@@ -927,6 +927,38 @@ public PersonalRelationshipType GetPersonalRelationshipTypeByDescription(string
...
+            var secondRelationshipDetails = _databaseContext.PersonalRelationshipDetails.FirstOrDefault(prd => prd.PersonalRelationshipId == secondRelationship.Id);
+
+            if (relationshipDetails != null)
+            {
+                _databaseContext.PersonalRelationshipDetails.Remove(relationshipDetails);
+            }
+            if (secondRelationshipDetails != null)
+            {
+                _databaseContext.PersonalRelationshipDetails.Remove(secondRelationshipDetails);
+            }
```
Comment (chubberlisk) [inline SocialCareCaseViewerApi/V1/Gateways/DatabaseGateway.cs:954]: Have we had a look to see if EF Core can help us out a bit here? Looks like we might be about to `Include` `PersonalRelationshipDetail` and then remove, that way we don't have to do two finds for them. https://docs.microsoft.com/en-us/ef/core/saving/cascade-delete#deleting-a-principalparent

File: SocialCareCaseViewerApi/V1/Gateways/DatabaseGateway.cs (hunk 4 above)
Comment (mapleleafjack) [reply SocialCareCaseViewerApi/V1/Gateways/DatabaseGateway.cs:936]: Test added :)

File: SocialCareCaseViewerApi.Tests/V1/UseCase/PersonalRelationships/PersonalRelationshipsExecuteDeleteUseCaseTests.cs [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+        [Test]
+        public void WhenPersonHasPersonalRelationshipsDeleteThem()
+        {
+            _mockDatabaseGateway.Setup(x => x.GetPersonalRelationshipById(It.IsAny<long>())).Returns(_relationship);
+
+            _personalRelationshipsUseCase.ExecuteDelete(_relationship.Id);
+
+            _mockDatabaseGateway.Verify(gateway => gateway.GetPersonalRelationshipById(_relationship.Id));
+            _mockDatabaseGateway.Verify(gateway => gateway.DeleteRelationships(_relationship));
+        }
```
Comment (chubberlisk) [inline SocialCareCaseViewerApi.Tests/V1/UseCase/PersonalRelationships/PersonalRelationshipsExecuteDeleteUseCaseTests.cs:66]: What do you think about splitting this test into two? One that shows that the use case will check that the ID provided exists as a personal relationship and then the other deletes them. This is so the tests clearly documents that the use case will validate the ID.

**Review by chubberlisk — APPROVED** [review]
🚢

**Pull request merged**